mash/mash.go

```diff
@@ -0,0 +1,92 @@
+/*
+Package mash is for sketching sequence data to make it easier to search for and against.
+
+The package is named mash after the mash sketching algorithm, which is based on the MinHash algorithm.
+
+Mash: fast genome and metagenome distance estimation using MinHash.
+Ondov, B.D., Treangen, T.J., Melsted, P. et al.
+Genome Biol 17, 132 (2016).
+https://doi.org/10.1186/s13059-016-0997-x
+
+Mash Screen: high-throughput sequence containment estimation for genome discovery.
+Ondov, B., Starrett, G., Sappington, A. et al.
+Genome Biol 20, 232 (2019).
+https://doi.org/10.1186/s13059-019-1841-x
+
+The idea is that we can sketch a sequence of data, and then compare the sketch to other sketches to see how similar they are.
+This saves a bunch of computation time and memory, because we don't have to compare the entire sequence to another sequence.
+
+TTFN,
+Tim
+*/
+package mash
+
+import "github.com/spaolacci/murmur3"
+
+// sketch algorithm
+// slide a window of size k along the sequence
+// for each kmer in the window, hash it to a 32 or 64 bit number
+// keep the minimum hash value of all the kmers in the window up to a given sketch size
+// the sketch is a vector of the minimum hash values
+
+// what are mash's inputs and outputs?
+// inputs: a
+
+type Mash struct {
+	KmerSize   uint
+	SketchSize uint
+	Sketches   []uint32
+}
+
+func NewMash(kmerSize uint, sketchSize uint) *Mash {
+	return &Mash{
+		KmerSize:   kmerSize,
+		SketchSize: sketchSize,
+		Sketches:   make([]uint32, sketchSize),
+	}
+}
+
+func (m *Mash) Sketch(sequence string) {
+	// slide a window of size k along the sequence
+	for kmerStart := 0; kmerStart < len(sequence)-int(m.KmerSize); kmerStart++ {
+		kmer := sequence[kmerStart : kmerStart+int(m.KmerSize)]
+		// hash the kmer to a 32 bit number
+		hash := murmur3.Sum32([]byte(kmer))
+		// keep the minimum hash value of all the kmers in the window up to a given sketch size
+		// the sketch is a vector of the minimum hash values
+		var biggestHashIndex int
+
+		// find the biggest hash value in the sketch
+		for i := 0; i < len(m.Sketches); i++ {
+			if m.Sketches[i] == 0 {
+				biggestHashIndex = i
+				break
+			} else if m.Sketches[i] > m.Sketches[biggestHashIndex] {
+				biggestHashIndex = i
+			}
+		}
+		m.Sketches[biggestHashIndex] = hash
+	}
+}
+
+// distance algorithm
+// compare the sketches of two sequences
+// count the number of hashes that are the same
+// divide the number of hashes that are the same by the total number of hashes
+// the result is the distance between the two sequences
+func (m *Mash) Distance(other *Mash) float64 {
```

Comment:
I personally think Distance should be a function (with two inputs) instead of a method, perhaps with a value instead of a pointer? Kinda like `strings.Contains()`

The reason is that I am imagining this being used across a LOT of data, and functional concurrency seems easier. I might be totally wrong here though - it could be that the performance gains with methods would be a lot better than what a purely functional approach would give you. 

Thoughts @carreter ?

Reply:
I don't think there is a performance difference between methods and functions, totally unsure though.

I also don't think this makes functional concurrency harder, as methods are still first-class objects. Could easily do the following:

```
var center *Mash // The Mash we will use as the center for distance calculations
var others []*Mash // The Mashes we are comparing center to
distances := make([]float64, len(otherMashes))

for i, other := range others {
    // Reassignments necessary as we are using loop variables in closures.
    i := i
    other := other

    go func() {
        distances[i] = center.Distance(other)
    }
}
```

Imagine that I did a waitgroup or something in the above code to ensure all the goroutines complete before we try and read from `distances`.

Reply:
Alternatively, imagine we had a `MapConcurrent` function:

```
func MapConcurrent[T any, O any](input []T, f func(T) O) []O {
     // Build output slice by applying function f to all elements of input, concurrently. 
}
```

Then you could simplify my previous example to this:

```
var center *Mash // The Mash we will use as the center for distance calculations
var others []*Mash // The Mashes we are comparing center to
distances := MapConcurrent(others, center.Distance)
```

Reply:
I'm ambivalent as to leaving this as a method vs turning it into a function. I'm assuming distances are commutative - if they aren't, then I would have a slight preference for keeping it as a method to make the directionality clearer (`m` to `other`).

Reply:
You're right, it shouldn't matter. If I am reading it right, distances are commutative. I'm going to resolve!